File: CSharp/ExtensibleCommands/ExtensibleCommands/ConditionalCommand.cs
using System;
using System.Collections.Generic;

namespace ExtensibleCommands
{
    /// <summary>
    /// Implements command branching based on a conditional boolean flag
    /// </summary>
    public class ConditionalCommand : Command
    {
        /// <summary> Command to run if the conditional flag is true </summary>
        public ICommand TrueCommand { get; }

        /// <summary> Command to run if the conditional flag is false </summary>
        public ICommand FalseCommand { get; }
    
        /// <summary> List of all child command objects (1st level only) </summary>
        public override IEnumerable<ICommand> Children
        {
            get
            {
                var children = new List<ICommand>();
                children.Add(TrueCommand);
                children.Add(FalseCommand);
                return children;
            }
        }

        /// <summary> List of all descendant command objects (all levels) </summary>
        public override IEnumerable<ICommand> Descendants
        {
            get
            {
                var descendants = new List<ICommand>();
                descendants.Add(TrueCommand);
                descendants.AddRange(TrueCommand.Descendants);
                descendants.Add(FalseCommand);
                descendants.AddRange(FalseCommand.Descendants);
                return descendants;
            }
        }

        /// <summary> Predicate to evaluate to decide which sub-command to run </summary>
        private readonly Func<bool> _predicate;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="predicate">Predicate to evaluate to decide which sub-command to run</param>
        /// <param name="trueCommand">Command to run if the conditional flag is true</param>
        /// <param name="falseCommand">Command to run if the conditional flag is false</param>
        public ConditionalCommand(Func<bool> predicate, ICommand trueCommand, ICommand falseCommand) 
            : this(predicate, trueCommand, falseCommand, "Conditional") { }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="predicate">Predicate to evaluate to decide which sub-command to run</param>
        /// <param name="trueCommand">Command to run if the conditional flag is true</param>
        /// <param name="falseCommand">Command to run if the conditional flag is false</param>
        /// <param name="name">Command name</param>
        public ConditionalCommand(Func<bool> predicate, ICommand trueCommand, ICommand falseCommand, string name)
        {
            if (predicate == null)
                throw new Exception(string.Format("Predicate is NULL in ConditionalCommand {0}", name));
            if (trueCommand == null)
                throw new Exception(string.Format("True Command is NULL in ConditionalCommand {0}", Name));
            if (falseCommand == null)
                throw new Exception(string.Format("False Command is NULL in ConditionalCommand {0}", Name));

            Name = name;
            _predicate = predicate;
            TrueCommand = trueCommand;
            FalseCommand = falseCommand;
        }

        protected override void Execute()
        {
            if (_predicate())
                TrueCommand.Run();
            else
                FalseCommand.Run();
        }

        /// <summary> Set the main command state based on the child commands states </summary>
        protected override void CheckErrors()
        {
            if (TrueCommand.CurrentState == State.Aborted ||
               FalseCommand.CurrentState == State.Aborted)
            {
                CurrentState = State.Aborted;
            }
            else if (TrueCommand.CurrentState == State.Failed)
            {
                CurrentState = State.Failed;
                Exception = TrueCommand.Exception;
            }
            else if (FalseCommand.CurrentState == State.Failed)
            {
                CurrentState = State.Failed;
                Exception = FalseCommand.Exception;
            }
            else if (TrueCommand.CurrentState == State.Completed ||
               FalseCommand.CurrentState == State.Completed)
            {
                CurrentState = State.Completed;
            }
        }
    }
}

File: Java/ExtensibleCommands/src/main/java/org/extensiblecommands/ConditionalCommand.java
package org.extensiblecommands;

import java.util.ArrayList;

import static java.util.Objects.requireNonNull;

/**
 * Implements command branching based on a conditional boolean flag
 */
public class ConditionalCommand extends AbstractCommand implements Command {
    /**
     * Command to run if the conditional flag is true
     */
    private final Command trueCommand;

    /**
     * Command to run if the conditional flag is false
     */
    private final Command falseCommand;

    /**
     * Predicate to evaluate to decide which sub-command to run
     */
    private final PredicateDelegate predicate;

    /**
     * Constructor
     * @param predicate         Predicate to evaluate to decide which sub-command to run
     * @param trueCommand       Command to run if the conditional flag is true
     * @param falseCommand      Command to run if the conditional flag is false
     */
    public ConditionalCommand(PredicateDelegate predicate, Command trueCommand, Command falseCommand) {
        this(predicate, trueCommand, falseCommand, "Conditional");
    }

    /**
     * Constructor
     * @param predicate         Predicate to evaluate to decide which sub-command to run
     * @param trueCommand       Command to run if the conditional flag is true
     * @param falseCommand      Command to run if the conditional flag is false
     * @param name              Command name
     */
    public ConditionalCommand(PredicateDelegate predicate, Command trueCommand, Command falseCommand, String name) {
        requireNonNull(predicate, String.format("Predicate is NULL in ConditionalCommand %s", name));
        requireNonNull(trueCommand, String.format("True Command is NULL in ConditionalCommand %s", name));
        requireNonNull(falseCommand, String.format("False Command is NULL in ConditionalCommand %s", name));

        this.name = name;
        this.predicate = predicate;
        this.trueCommand = trueCommand;
        this.falseCommand = falseCommand;
    }

    /**
     * @return      Command to run if the conditional flag is true
     */
    public final Command getTrueCommand() {
        return trueCommand;
    }

    /**
     * @return      Command to run if the conditional flag is false
     */
    public final Command getFalseCommand() {
        return falseCommand;
    }

    /**
     * @return      List of all child command objects (1st level only)
     */
    @Override
    public Iterable<Command> getChildren() {
        var children = new ArrayList<Command>();
        children.add(trueCommand);
        children.add(falseCommand);
        return children;
    }

    /**
     * @return      List of all descendant command objects (all levels)
     */
    @Override
    public Iterable<Command> getDescendants() {
        var descendants = new ArrayList<Command>();
        descendants.add(trueCommand);
        trueCommand.getDescendants().forEach(c -> descendants.add(c));
        descendants.add(falseCommand);
        falseCommand.getDescendants().forEach(c -> descendants.add(c));
        return descendants;
    }

    @Override
    protected void execute() throws Exception {
        if (predicate.evaluateCondition())
            trueCommand.run();
        else
            falseCommand.run();
    }

    /**
     * Set the main command state based on the child commands states
     */
    @Override
    protected void checkErrors() {
        if (trueCommand.getState() == State.Aborted ||
                falseCommand.getState() == State.Aborted) {
            setState(State.Aborted);
        }
        else if (trueCommand.getState() == State.Failed) {
            setState(State.Failed);
            exception = trueCommand.getException();
        }
        else if (falseCommand.getState() == State.Failed) {
            setState(State.Failed);
            exception = falseCommand.getException();
        }
        else if (trueCommand.getState() == State.Completed ||
                falseCommand.getState() == State.Completed) {
            setState(State.Completed);
        }
    }
}
